commit bdb596ddc07596fa212f231135b87d0b9178f6f8
Author: Michal Vasko <mvasko@cesnet.cz>
Date:   Mon Apr 29 08:59:30 2019 +0200

    resolve BUGFIX handle yet unresolved if-features
    
    Fixes #772

diff --git a/src/resolve.c b/src/resolve.c
index e8bd7576..af83bdfb 100644
--- a/src/resolve.c
+++ b/src/resolve.c
@@ -1402,7 +1402,7 @@ resolve_iffeature(struct lys_iffeature *expr)
 {
     int index_e = 0, index_f = 0;
 
-    if (expr->expr) {
+    if (expr->expr && expr->features[0]) {
         return resolve_iffeature_recursive(expr, &index_e, &index_f);
     }
     return 0;
