commit fe3066122ef72c82415811d25e9e3fad622c0a99
Author: Cristy <urban-warrior@imagemagick.org>
Date:   Tue Jun 18 11:51:46 2019 -0400

    https://github.com/ImageMagick/ImageMagick/issues/1600

diff --git a/coders/bmp.c b/coders/bmp.c
index fdabac950..c964834fe 100644
--- a/coders/bmp.c
+++ b/coders/bmp.c
@@ -1506,7 +1506,7 @@ static Image *ReadBMPImage(const ImageInfo *image_info,ExceptionInfo *exception)
         if (GetNextImageInList(image) == (Image *) NULL)
           {
             status=MagickFalse;
-            return((Image *) NULL);
+            break;
           }
         image=SyncNextImageInList(image);
         status=SetImageProgress(image,LoadImagesTag,TellBlob(image),
diff --git a/coders/viff.c b/coders/viff.c
index 5f89ad15e..05aac7c55 100644
--- a/coders/viff.c
+++ b/coders/viff.c
@@ -773,8 +773,8 @@ static Image *ReadVIFFImage(const ImageInfo *image_info,
         AcquireNextImage(image_info,image,exception);
         if (GetNextImageInList(image) == (Image *) NULL)
           {
-            image=DestroyImageList(image);
-            return((Image *) NULL);
+            status=MagickFalse;
+            break;
           }
         image=SyncNextImageInList(image);
         status=SetImageProgress(image,LoadImagesTag,TellBlob(image),
