commit cb7190cd2c691fd93e4d3664f3fce6c19ee001dd	cb7190cd2c691fd93e4d3664f3fce6c19ee001dd
Author: Janne Grunau <janne-libav@jannau.net>
Date:   Fri Mar 23 22:30:38 2012 +0100

    rv34: error out on size changes with frame threading

diff --git a/libavcodec/rv34.c b/libavcodec/rv34.c
index f9ea40d6c8..da5d437b07 100644
--- a/libavcodec/rv34.c
+++ b/libavcodec/rv34.c
@@ -1662,6 +1662,13 @@ int ff_rv34_decode_frame(AVCodecContext *avctx,
         if (s->width != si.width || s->height != si.height) {
             int err;
 
+            if (HAVE_THREADS &&
+                (s->avctx->active_thread_type & FF_THREAD_FRAME)) {
+                av_log_missing_feature(s->avctx, "Width/height changing with "
+                                       "frame threading is", 0);
+                return AVERROR_PATCHWELCOME;
+            }
+
             av_log(s->avctx, AV_LOG_WARNING, "Changing dimensions to %dx%d\n",
                    si.width, si.height);
             ff_MPV_common_end(s);
