commit 68154c05cf40a80b6f2e2dd9fdc4428570f875f0
Author: Cristy <urban-warrior@imagemagick.org>
Date:   Thu Nov 19 18:40:10 2020 +0000

    fix shell injection vulnerability via the -authenticate option

diff --git a/coders/pdf.c b/coders/pdf.c
index 3a14124fc..224a5da47 100644
--- a/coders/pdf.c
+++ b/coders/pdf.c
@@ -591,7 +591,8 @@ static Image *ReadPDFImage(const ImageInfo *image_info,ExceptionInfo *exception)
     (void) ConcatenateMagickString(options,"-dInterpolateControl=-1 ",
       MagickPathExtent);
   option=GetImageOption(image_info,"authenticate");
-  if ((option != (char *) NULL) && (strpbrk(option,"&;<>|\"") == (char *) NULL))
+  if ((option != (char *) NULL) &&
+      (strpbrk(option,"&;<>|\"'") == (char *) NULL))
     {
       char
         passphrase[MagickPathExtent];
