commit 11f8838bf99ea0a6f0b76f9760c43704d00c4ff6
Author: Sebastian Pipping <sebastian@pipping.org>
Date:   Wed Jun 12 15:42:22 2019 +0200

    xmlparse.c: Fix extraction of namespace prefix from XML name (#186)

diff --git a/expat/lib/xmlparse.c b/expat/lib/xmlparse.c
index 30d55c5c..737d7cd2 100644
--- a/expat/lib/xmlparse.c
+++ b/expat/lib/xmlparse.c
@@ -6071,7 +6071,7 @@ setElementTypePrefix(XML_Parser parser, ELEMENT_TYPE *elementType)
       else
         poolDiscard(&dtd->pool);
       elementType->prefix = prefix;
-
+      break;
     }
   }
   return 1;
