commit b831d9051105226ceb874279276e395abd26c37d
Author: Cristy <urban-warrior@imagemagick.org>
Date:   Sat May 7 08:23:15 2016 -0400

    Remove support for internal ephemeral coder.

diff --git a/MagickCore/image.c b/MagickCore/image.c
index 7be95a381..1b3519571 100644
--- a/MagickCore/image.c
+++ b/MagickCore/image.c
@@ -2614,7 +2614,6 @@ MagickExport MagickBooleanType SetImageInfo(ImageInfo *image_info,
           "BROWSE",
           "DCRAW",
           "EDIT",
-          "EPHEMERAL",
           "LAUNCH",
           "MPEG:DECODE",
           "MPEG:ENCODE",
@@ -2679,10 +2678,7 @@ MagickExport MagickBooleanType SetImageInfo(ImageInfo *image_info,
       if (IsMagickConflict(magic) == MagickFalse)
         {
           (void) CopyMagickString(image_info->magick,magic,MagickPathExtent);
-          if (LocaleCompare(magic,"EPHEMERAL") != 0)
-            image_info->affirm=MagickTrue;
-          else
-            image_info->temporary=MagickTrue;
+          image_info->affirm=MagickTrue;
         }
     }
   magick_info=GetMagickInfo(magic,sans_exception);
diff --git a/MagickCore/magick.c b/MagickCore/magick.c
index c62ef307b..6dcc463ab 100644
--- a/MagickCore/magick.c
+++ b/MagickCore/magick.c
@@ -1009,12 +1009,6 @@ static MagickBooleanType IsMagickTreeInstantiated(ExceptionInfo *exception)
           if (magick_list == (SplayTreeInfo *) NULL)
             ThrowFatalException(ResourceLimitFatalError,
               "MemoryAllocationFailed");
-          magick_info=AcquireMagickInfo("EPHEMERAL","EPHEMERAL","Internal format");
-          magick_info->flags|=CoderStealthFlag;
-          status=AddValueToSplayTree(magick_list,magick_info->name,magick_info);
-          if (status == MagickFalse)
-            ThrowFatalException(ResourceLimitFatalError,
-              "MemoryAllocationFailed");
 #if defined(MAGICKCORE_MODULES_SUPPORT)
           (void) GetModuleInfo((char *) NULL,exception);
 #endif
