commit 31fa3304049fc406a201a72293cce140f0557dca
Author: Yukihiro "Matz" Matsumoto <matz@ruby.or.jp>
Date:   Sun Jan 16 22:22:46 2022 +0900

    class.c: add `obj->c` check before `prepare_singleton_class()`.

diff --git a/src/class.c b/src/class.c
index ec9c648e5..8751dcaf0 100644
--- a/src/class.c
+++ b/src/class.c
@@ -357,6 +357,7 @@ prepare_singleton_class(mrb_state *mrb, struct RBasic *o)
 {
   struct RClass *sc, *c;
 
+  mrb_assert(o->c);
   if (o->c->tt == MRB_TT_SCLASS) return;
   sc = MRB_OBJ_ALLOC(mrb, MRB_TT_SCLASS, mrb->class_class);
   sc->flags |= MRB_FL_CLASS_IS_INHERITED;
@@ -1682,6 +1683,7 @@ mrb_singleton_class_ptr(mrb_state *mrb, mrb_value v)
     break;
   }
   obj = mrb_basic_ptr(v);
+  if (obj->c == NULL) return NULL;
   prepare_singleton_class(mrb, obj);
   return obj->c;
 }
