commit 8c3166e1c302c3ba80d9742ae46161c0fa8e2606
Author: Paul B Mahol <onemda@gmail.com>
Date:   Sun Oct 13 23:21:35 2019 +0200

    avfilter/vf_bwdif: fix heap-buffer overflow
    
    Fixes #8261

diff --git a/libavfilter/vf_bwdif.c b/libavfilter/vf_bwdif.c
index 37165584cf..b6aed7a450 100644
--- a/libavfilter/vf_bwdif.c
+++ b/libavfilter/vf_bwdif.c
@@ -343,8 +343,8 @@ static int config_props(AVFilterLink *link)
     if(yadif->mode&1)
         link->frame_rate = av_mul_q(link->src->inputs[0]->frame_rate, (AVRational){2,1});
 
-    if (link->w < 3 || link->h < 3) {
-        av_log(ctx, AV_LOG_ERROR, "Video of less than 3 columns or lines is not supported\n");
+    if (link->w < 3 || link->h < 4) {
+        av_log(ctx, AV_LOG_ERROR, "Video of less than 3 columns or 4 lines is not supported\n");
         return AVERROR(EINVAL);
     }
 
