commit a1098f81c252b317ad34ea978aea2bc47760b215
Author: Krzysztof Stachowiak <krzysiek.stachowiak@gmail.com>
Date:   Tue Mar 13 11:28:49 2018 +0100

    Add bounds check before signature length read

diff --git a/library/ssl_cli.c b/library/ssl_cli.c
index 2534346a4..279a127ba 100644
--- a/library/ssl_cli.c
+++ b/library/ssl_cli.c
@@ -2478,6 +2478,14 @@ static int ssl_parse_server_key_exchange( mbedtls_ssl_context *ssl )
         /*
          * Read signature
          */
+
+        if( p > end - 2 )
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad server key exchange message" ) );
+            mbedtls_ssl_send_alert_message( ssl, MBEDTLS_SSL_ALERT_LEVEL_FATAL,
+                                            MBEDTLS_SSL_ALERT_MSG_DECODE_ERROR );
+            return( MBEDTLS_ERR_SSL_BAD_HS_SERVER_KEY_EXCHANGE );
+        }
         sig_len = ( p[0] << 8 ) | p[1];
         p += 2;
 
