commit e14fd0a2801f73bdc123baf4fbab97dec55919eb
Author: dirk <dirk@git.imagemagick.org>
Date:   Fri Jan 15 01:27:31 2016 +0100

    Added missing call to ConstrainColormapIndex.

diff --git a/coders/psd.c b/coders/psd.c
index 5f5fba6c6..5a5c63380 100644
--- a/coders/psd.c
+++ b/coders/psd.c
@@ -835,7 +835,8 @@ static MagickBooleanType ReadPSDChannelPixels(Image *image,
                   SetPixelIndex(image,(((unsigned char) pixel) &
                     (0x01 << (7-bit))) != 0 ? 0 : 255,q);
                   SetPixelViaPixelInfo(image,image->colormap+(ssize_t)
-                    GetPixelIndex(image,q),q);
+                    ConstrainColormapIndex(image,GetPixelIndex(image,q),
+                      exception),q);
                   q+=GetPixelChannels(image);
                   x++;
                 }
