commit ff2431d8f17d4a7c906438042a649b04aec93558
Author: Dirk Lemstra <dirk@git.imagemagick.org>
Date:   Sun May 7 11:13:07 2017 +0200

    Replace image in list to fix issue reported in #467.

diff --git a/coders/bmp.c b/coders/bmp.c
index cc036a1f5..207118ed1 100644
--- a/coders/bmp.c
+++ b/coders/bmp.c
@@ -1380,7 +1380,7 @@ static Image *ReadBMPImage(const ImageInfo *image_info,ExceptionInfo *exception)
         if (flipped_image != (Image *) NULL)
           {
             DuplicateBlob(flipped_image,image);
-            image=DestroyImage(image);
+            ReplaceImageInList(&image, flipped_image);
             image=flipped_image;
           }
       }
