commit 916d7bbd2c66a286d379dbd94bc6035c8fab937c
Author: Dirk Lemstra <dirk@lemstra.org>
Date:   Wed Nov 27 14:13:54 2019 +0100

    Removed invalid free reported in #1791.

diff --git a/coders/png.c b/coders/png.c
index defb0d37b..853dd7f4a 100644
--- a/coders/png.c
+++ b/coders/png.c
@@ -7677,10 +7677,7 @@ static Image *ReadOneMNGImage(MngInfo* mng_info, const ImageInfo *image_info,
       next_image=CoalesceImages(image,exception);
       image=DestroyImageList(image);
       if (next_image == (Image *) NULL)
-        {
-          mng_info=MngInfoFreeStruct(mng_info);
-          return((Image *) NULL);
-        }
+        return((Image *) NULL);
       image=next_image;
 
       for (next=image; next != (Image *) NULL; next=next_image)
